File: Solves/3xxx/30xx/3043.go
package main

import (
	"strconv"
)

func countDigits(num int) int {
	return len(strconv.Itoa(num))
}

func longestCommonPrefix(arr1 []int, arr2 []int) int {
	var ans int
	pref := make(map[int]int)
	for _, num := range arr1 {
		for num > 0 {
			pref[num]++
			num /= 10
		}
	}

	for _, num := range arr2 {
		for num > 0 {

			if pref[num] > 0 {
				ans = max(ans, countDigits(num))
				break
			}
			num /= 10
		}
	}
	return ans
}
func main() {
	arr1 := []int{1, 10, 100}
	arr2 := []int{1000}
	println(longestCommonPrefix(arr1, arr2)) // Output: 1

	arr1 = []int{1, 2, 3}
	arr2 = []int{4, 4, 4}
	println(longestCommonPrefix(arr1, arr2)) // Output: 0
}

File: Solves/3xxx/30xx/3043.kt
class TrieNode{
    val children = mutableMapOf<Char, TrieNode>()
}

internal class Trie{
    private val root = TrieNode()
    
    fun addNum(num : String)
    {
        var currentNode = root
        for(c in num)
        {
            if(!currentNode.children.containsKey(c))
            {
                currentNode.children[c] = TrieNode()
            }

            currentNode = currentNode.children[c]!!
        }
    }

    fun getPrefixSize(num: String): Int{
        var ans = 0

        var currentNode = root
        for(c in num)
        {
            if(!currentNode.children.containsKey(c)) break;
            currentNode = currentNode.children[c]!!;
            ans ++
        }

        return ans
    }
}

class Solution {
    fun longestCommonPrefix(arr1: IntArray, arr2: IntArray): Int {
        val trie = Trie()
        var res = 0
        
        for(num in arr1) {
            val numString = num.toString()
            trie.addNum(numString)
        }

        for(num in arr2) {
            val numString = num.toString()
            res = maxOf(res, trie.getPrefixSize(numString))
        }
        
        return res
    }
}